sale_margin_percentage/tests/test_sale.py

```diff
@@ -0,0 +1,164 @@
+import time
+from odoo.tests.common import TransactionCase
+from odoo.tests import tagged
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT
+
+
+@tagged("sale_order")
+class TestSaleOrder(TransactionCase):
+    """Test cases for sale.order model"""
+
+    def setUp(self):
+        super(TestSaleOrder, self).setUp()
+        self.date_now = time.strftime(DEFAULT_SERVER_DATE_FORMAT)
+        self.partner_camp = self.env.ref('base.res_partner_12')
+        self.prod_ipadmini = self.env.ref('product.product_product_6')
+        self.demo_user = self.env.ref('base.user_demo')
+        self.sale_order_id = self.env['sale.order'].create({
+            'partner_id': self.partner_camp.id,
+            'order_line': [(0, 0, {
+                'name': 'iPadMini',
+                'product_id': self.prod_ipadmini.id,
+                'product_uom': self.prod_ipadmini.uom_id.id,
+                'product_uom_qty': 1,
+            })],
+        })
+        self.sale_order_id.user_id = self.demo_user.id
+
+    def test_01_margin_cost_0(self):
+        """ Margin and Margin Percentage when cost is 0.0
+        """
+        with self.env.do_in_onchange():
+            product = self.prod_ipadmini
+            product.standard_price = 0.0
+            sale = self.env['sale.order'].create({
+                'partner_id': self.partner_camp.id,
+                'order_line': [(0, 0, {
+                    'name': 'iPadMini',
+                    'product_id': product.id,
+                    'product_uom': product.uom_id.id,
+                    'product_uom_qty': 4,
+                    'purchase_price': 0.0,
+                    'price_unit': 12.0,
+                })],
+            })
+            sale_line = sale.order_line
+            self.assertEqual(sale_line.margin, 48)
+            self.assertEqual(sale_line.margin_percentage, 100.0)
+
+    def test_02_margin_sale_price_0(self):
+        """ Margin and Margin Percentage when sale price is 0.0
+        """
+        with self.env.do_in_onchange():
```

Comment:
Use the Form class, please, [here](https://github.com/Vauxoo/addons-vauxoo/blob/8c9a1597b4e3bd3904ceca75ecd8cef24c7862ec/purchase_third_validation/tests/test_purchase_third_validation.py#L30) an example.

This comment is applicable for the below tests

Reply:
@hugho-ad AFAIK you helped @angelJuaGar to use the Form class, but it's not in the tests. Did you decided to don't use it?

Reply:
@angelJuaGar 

which problem did you get using form?

I remember you told me that onchange function didn't be called right?

Reply:
It is right. 
Using the form class, it does not detect the onchange function, applying a different process in the calculation of the percentage and causing different results than those expected in the test.

Reply:
It should be quite the opposite, the Form class is used in order to get onchanges triggered automatically, instead of calling them manually. Actually, using the Form class, the result should be closer to the functional tests.

Reply:
I understood that when it was recommended to me to use the form class but when applying it in the tests it was not reflected that the onchange was activated.
The test was replicated in a functional way to be able to check the results and they were incorrect.

Reply:
Hi @luisg123v
Is it okay to handle the test this way?

Reply:
@angelJuaGar not ideal, but it was that way, it's working as it is and I understand make it to work using the Form class could be a little bit complex, so It's OK for me.